frontend/dockerfile/shell/lex.go

```diff
@@ -472,3 +490,88 @@ func BuildEnvs(env []string) map[string]string {
 
 	return envs
 }
+
+// convert a shell-like wildcard pattern (? is a single char, * either the
+// shortest or longest (greedy) string) to an equivalent regular expression
+func convertShellPatternToRegex(pattern string, greedy bool) (*regexp.Regexp, error) {
+	var s scanner.Scanner
+	s.Init(strings.NewReader(pattern))
+	out := strings.Builder{}
+	out.Grow(len(pattern) + 4)
+
+	// match only at the beginning of the string
+	out.WriteByte('^')
+
+	// default: non-greedy wildcards
+	starPattern := ".*?"
+	if greedy {
+		starPattern = ".*"
+	}
+
+	for tok := s.Next(); tok != scanner.EOF; tok = s.Next() {
+		switch tok {
+		case '*':
+			out.WriteString(starPattern)
+			continue
+		case '?':
+			out.WriteByte('.')
+			continue
+		case '\\':
+			// } as part of ${} needs to be escaped, but the escape isn't part
+			// of the pattern
+			if s.Peek() == '}' {
+				continue
+			}
+			out.WriteRune('\\')
+			tok = s.Next()
+			if tok != '*' && tok != '?' && tok != '\\' {
+				return nil, errors.Errorf("invalid escape '\\%c'", tok)
+			}
+		// regex characters that need to be escaped
+		case '[', '{', '.', '+', '(', ')', '|', '^', '$':
+			out.WriteByte('\\')
```

Comment:
The main issue with `QuoteMeta` is that is escapes the `\}` again, so rather than double-unescaping it afterwards I added the loop above. The marked list contains the characters from the link (except `]`, for which escaping is optional, and `\`, `*`, `}`).